flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSourceReader.java

```diff
@@ -0,0 +1,213 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.base.source.hybrid;
+
+import org.apache.flink.api.connector.source.ReaderOutput;
+import org.apache.flink.api.connector.source.SourceEvent;
+import org.apache.flink.api.connector.source.SourceReader;
+import org.apache.flink.api.connector.source.SourceReaderContext;
+import org.apache.flink.api.connector.source.SourceSplit;
+import org.apache.flink.core.io.InputStatus;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Hybrid source reader that delegates to the actual source reader.
+ *
+ * <p>This reader is setup with a sequence of underlying source readers. At a given point in time,
+ * one of these readers is active. Underlying readers are opened and closed on demand as determined
+ * by the enumerator, which selects the active reader via {@link SwitchSourceEvent}.
+ *
+ * <p>When the underlying reader has consumed all input, {@link HybridSourceReader} sends {@link
+ * SourceReaderFinishedEvent} to the coordinator and waits for the {@link SwitchSourceEvent}.
+ */
+public class HybridSourceReader<T, SplitT extends SourceSplit>
+        implements SourceReader<T, HybridSourceSplit<SplitT>> {
+    private static final Logger LOG = LoggerFactory.getLogger(HybridSourceReader.class);
+    // Controls the interval at which to indicate to the coordinator that the current reader
+    // has consumed all input and the next source can be activated. This is necessary to not flood
+    // the coordinator with duplicate events.
+    private static final int SOURCE_READER_FINISHED_EVENT_DELAY = 250;
+    private SourceReaderContext readerContext;
+    private List<SourceReader<?, SplitT>> realReaders;
+    private int currentSourceIndex = -1;
+    private long lastCheckpointId = -1;
+    private SourceReader<?, SplitT> currentReader;
+    private long lastReaderFinishedMs;
+
+    public HybridSourceReader(
+            SourceReaderContext readerContext, List<SourceReader<?, SplitT>> readers) {
+        this.readerContext = readerContext;
+        this.realReaders = readers;
+    }
+
+    @Override
+    public void start() {
+        setCurrentReader(0);
+    }
+
+    @Override
+    public InputStatus pollNext(ReaderOutput output) throws Exception {
+        InputStatus status = currentReader.pollNext(output);
+        if (status == InputStatus.END_OF_INPUT) {
+            // trap END_OF_INPUT if this wasn't the final reader
+            LOG.debug(
```

Comment:
nit: this is an important and infrequent reader event, maybe `info` level logging

Reply:
It's related to the jitter question. Please take a look at the comment that I added. The important part is that this repeats in a tight loop.

Reply:
I see. I misunderstood earlier. Yeah, `MORE_AVAILABLE` will cause the tight loop. `NOTHING_AVAILABLE` is probably the right choice.

Reply:
I made the change to `NOTHING_AVAILABLE` and that appears to address the issue. The reader poll will be suspended until the previous future is marked complete, which occurs upon source switch.